# Sac à dos

### Le problème à résoudre

Nous disposons d’une clé USB qui est déjà bien remplie et sur laquelle il ne reste que 5 Go de libre. Nous souhaitons copier sur cette clé des fichiers vidéos pour l’emporter en voyage. Chaque fichier a un poids et chaque vidéo a une durée. La durée n’est pas proportionnelle au poids car les fichiers sont de format différents, certaines vidéos sont de grande qualité, d’autres sont très compressées.

Le tableau qui suit présente les fichiers disponibles avec les durées données en minutes 

<table>
<tr><td>Fichier</td><td>video 1</td><td>video 2</td><td>video 3</td><td>video 4</td><td>video 5</td><td>video 6</td><td>video 7</td></tr>
<tr><td>Durée</td><td>114</td><td>32</td><td>20</td><td>4</td><td>18</td><td>80</td><td>5</td></tr>
<tr><td>Poids</td><td>4.57 Go</td><td>630 Mo</td><td>1.65 Go</td><td>85 Mo</td><td>2.15 Go</td><td>2.71 Go</td><td>320 Mo</td></tr>
</table>

## Question : quel sera le choix le plus optimal ?

<ol><li>force brute</li>
    <li>algorithme glouton
        <ul>
            <li>par durée</li>
            <li>par poids</li>
            <li>par durée/poids</li>
        </ul>
    </li>
</ol>


In [26]:
# Jeu de données constitué à partir du tableau

videos = [['Video 1', 114, 4.57], ['Video 2', 32, 0.630],
['Video 3', 20, 1.65], ['Video 4', 4, 0.085],
['Video 5', 18, 2.15], ['Video 6', 80, 2.71],
['Video 7', 5, 0.320]]

In [27]:
def gain_duree(liste):
    valeur=0
    for elem in liste:
        valeur+=elem[1]
    return valeur

exemple = [['Video 1', 114, 4.57], ['Video 2', 32, 0.630],['Video 6', 80, 2.71] ]        
gain_duree(exemple)

226

In [28]:
def poids(liste):
    gigaoctet=0
    for elem in liste:
        gigaoctet+=elem[2]
    return gigaoctet

exemple = [['Video 1', 114, 4.57], ['Video 2', 32, 0.630],['Video 6', 80, 2.71] ]        
poids(exemple)

7.91

# Force brute

In [29]:
#Réponse

from itertools import combinations

def combinaison_optimiser(liste):
    
    duree_max=0 # initiatilisation de la valeur max
    poids_max=0  # initialisation du poids max
    comb_max=()  # initiatilisation de la combinaison max 
    
    for nbre in range(1,len(liste)):
        for comb in combinations ( liste , nbre) :
            # toutes les combinaisons 1, 2, 3, 4, 5, 6
            poids_comb=poids(comb)  # calcul du poids de la combinaison
            duree_comb=gain_duree(comb) # calcul de la valeur de la combinaison
            
            if poids_comb <= 5 and duree_max<duree_comb:
                duree_max=duree_comb # garde la valeur max
                comb_max=comb          # garde la meilleure combinaison
                poids_max=poids_comb
                
    return comb_max, duree_max, poids_max
   
print(combinaison_optimiser(videos))

((['Video 2', 32, 0.63], ['Video 3', 20, 1.65], ['Video 6', 80, 2.71]), 132, 4.99)


# Algorithme glouton -> duree(valeur)

le plus de musique par rapport à la durée

In [30]:
def glouton(liste, poids_max):
    
    liste_optimisee = []
    stockage_dispo=poids_max
    
    # tri décroissant de la liste en entrée par rapport à la valeur
    liste = sorted(liste, key=lambda liste : liste[1], reverse = True)
        
    for elem in liste:
        if (stockage_dispo - elem[2])>=0:
            liste_optimisee.append(elem)
            stockage_dispo-=elem[2]
    
    poids_liste=poids(liste_optimisee )  # calcul du poids de la combinaison
    duree_liste=gain_duree(liste_optimisee ) # calcul de la valeur de la combinaison
    
    return liste_optimisee, duree_liste, poids_liste    
        
        
print(glouton(videos, 5))


([['Video 1', 114, 4.57], ['Video 7', 5, 0.32], ['Video 4', 4, 0.085]], 123, 4.9750000000000005)


# Algorithme glouton -> poids

le plus de musique par rapport au poids

In [34]:
def glouton(liste, poids_max):
    
    liste_optimisee = []
    stockage_dispo=poids_max
    
    # tri croissant de la liste en entrée par rapport au poids
    liste = sorted(liste, key=lambda liste : liste[2], reverse = False)
    for elem in liste:
        if (stockage_dispo - elem[2])>=0:
            liste_optimisee.append(elem)
            stockage_dispo-=elem[2]
    
    poids_liste=poids(liste_optimisee )  # calcul du poids de la combinaison
    duree_liste=gain_duree(liste_optimisee ) # calcul de la valeur de la combinaison
    
    return liste_optimisee, duree_liste, poids_liste    
        
        
print(glouton(videos, 5))


([['Video 4', 4, 0.085], ['Video 7', 5, 0.32], ['Video 2', 32, 0.63], ['Video 3', 20, 1.65], ['Video 5', 18, 2.15]], 79, 4.835)


# Algorithme glouton -> rapport duree / poids

In [19]:
def ajout_rapport_duree_poids(liste):
    list
    for elem in liste:
        rapport=round(elem[1]/elem[2],2)
        elem.append(rapport)
    return liste

print(ajout_rapport_duree_poids(videos))

[['Video 1', 114, 4.57, 24.95], ['Video 2', 32, 0.63, 50.79], ['Video 3', 20, 1.65, 12.12], ['Video 4', 4, 0.085, 47.06], ['Video 5', 18, 2.15, 8.37], ['Video 6', 80, 2.71, 29.52], ['Video 7', 5, 0.32, 15.62]]


In [23]:
def glouton_duree_poids(liste, poids_max):
    
    liste_optimisee = []
    stockage_dispo=poids_max
    
    # tri décroissant de la liste en entrée par rapport à la valeur/poids
    liste = sorted(liste, key=lambda liste : liste[3], reverse = True)
        
    for elem in liste:
        if (stockage_dispo - elem[2])>=0:
            liste_optimisee.append(elem)
            stockage_dispo-=elem[2]
    
    poids_liste=poids(liste_optimisee )  # calcul du poids de la combinaison
    duree_liste=gain_duree(liste_optimisee ) # calcul de la valeur de la combinaison
    
    return liste_optimisee, duree_liste, poids_liste    
        
        
print(glouton_duree_poids(videos, 5))

([['Video 2', 32, 0.63, 50.79], ['Video 4', 4, 0.085, 47.06], ['Video 6', 80, 2.71, 29.52], ['Video 7', 5, 0.32, 15.62]], 121, 3.7449999999999997)


# Conclusion

force brute : (['Video 2', 32, 0.63], ['Video 3', 20, 1.65], ['Video 6', 80, 2.71]), 132, 4.99

glouton durée : [['Video 1', 114, 4.57], ['Video 7', 5, 0.32], ['Video 4', 4, 0.085]], 123, 4.9750000000000005

glouton poids : [['Video 4', 4, 0.085], ['Video 7', 5, 0.32], ['Video 2', 32, 0.63], ['Video 3', 20, 1.65], ['Video 5', 18, 2.15]], 79, 4.835

glouton rapport : [['Video 2', 32, 0.63, 50.79], ['Video 4', 4, 0.085, 47.06], ['Video 6', 80, 2.71, 29.52], ['Video 7', 5, 0.32, 15.62]], 121, 3.7449999999999997

Nous constatons que le critère de durée est le plus intéressant, puisque la valeur totale est 123. Cette solution n’est cependant pas la solution optimale trouvée par force brute ['Video 2', 'Video 3', 'Video 6'] 132